File: Basket/src/com/lyonsys/basket/Basket.java
package com.lyonsys.basket;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;


public class Basket<T extends Product> {

    private final ConcurrentMap<T, PriceQuantity> items = new ConcurrentHashMap<>();
    private Map<T, Double> productPriceMap;

   public Basket(Map<T, Double> productPriceMap) {
        this.productPriceMap = productPriceMap;
    }


    public void addToBasket(T product, int quantity) {
        validateQuantity(quantity);
        if (this.productPriceMap == null || this.productPriceMap.get(product) == null) {
            throw new BasketException("No price found for product " + product.getProductId());
        }
        if (this.items.get(product) == null) {
            this.items.putIfAbsent(product, new PriceQuantity(productPriceMap.get(product), Integer.valueOf(0)));
        }
        PriceQuantity priceQuantity = items.get(product);
        synchronized (priceQuantity) {
            priceQuantity.setQuantity(priceQuantity.getQuantity() + quantity);
        }
    }

    public PriceQuantity removeFromBasket(Product product) {
        return this.items.remove(product);
    }

    public void updateQuantity(Product product, int newQuantity) {
        validateQuantity(newQuantity);
       this.updatePriceQuantity(product, null, newQuantity);
    }

    public void updatePrice(Product product,double price) {
        this.updatePriceQuantity(product, price, null);
    }

    public void updatePriceQuantity(Product product, Double newPrice, Integer newQuantity) {
        PriceQuantity priceQuantity = items.get(product);
        if (priceQuantity != null) {
            synchronized (priceQuantity) {
                if (newPrice != null) {
                    priceQuantity.setPrice(newPrice);
                }
                if (newQuantity != null) {
                    validateQuantity(newQuantity);
                    priceQuantity.setQuantity(newQuantity);
                }
            }
        }
    }

    private void validateQuantity(int quantity)
    {
        if (quantity <= 0 )
        {
            throw new BasketException("Quantity <" + quantity + "> is invalid");
        }
    }
    public double getTotalCost() {
        final DoubleValueHolder total = new DoubleValueHolder();

        items.keySet().stream().forEach(q -> {
            total.add(items.get(q).getPrice() * items.get(q).getQuantity());
        });
        return total.getValue();
    }

    private static class DoubleValueHolder {
        public double getValue() {
            return value;
        }

        double value;

        public void add(double delta)
        {
            value += delta;
        }
    }

}

File: scala/Basket/src/com/lyonsys/basket/Basket.scala
package com.lyonsys.basket

import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.ConcurrentHashMap;
import scala.collection._;
import scala.collection.convert.decorateAsScala._;

class Basket {

  private var items: concurrent.Map[Product, AtomicInteger] = new ConcurrentHashMap().asScala;
  private var prices = scala.collection.mutable.Map[Product, Double]();

  private val fruit = Set("Apple", "Orange", "Peach", "Banana", "Lemon");

  def initPrice(item: Product, price: Double): Unit =
    {
      prices.put(item, price);
    }

  def addToBasket(item: Product, quantity: Int) =
    {
      validateBasketItem(item);
      validateQuantty(quantity);
      if (!this.items.contains(item)) {
        this.items.putIfAbsent(item, new AtomicInteger(0));
      }
      this.items.get(item).get.getAndAdd(quantity);
    }
  
    def removeFromBasket(item: Product) =
    {
      items.remove(item);
    }

  def getTotalCost(): Double =
    {
      var totalCost: Double = 0;
      this.items.foreach(f => totalCost += prices.get(f._1).get.doubleValue() * f._2.get.doubleValue());
      return totalCost;
    }

  private def validateBasketItem(item: Product): Unit =
    {
      if (item == null || !fruit.contains(item.name)) {
        throw new RuntimeException("Invalid or unsupported product");
      }
    }
  private def validateQuantty(quantity: Int): Unit =
    {
      if (quantity <= 0) {
        throw new RuntimeException("Invalid quantity");
      }
    }

}